<a href="https://colab.research.google.com/github/aarakova/VvAD/blob/lab1/%D0%9B%D0%A01_%D0%A0%D0%B0%D0%BA%D0%BE%D0%B2%D0%B0_15.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Цель работы**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

**Номер варианта**

15 вариант.
Данные содержат информацию о кредитных заявках:
1.	возраст
2.	доход
3.	статус домовладения (Собственный, Ипотека, Аренда)
4.	стаж работы в годах
5.	цель кредита
6.	сумма кредита по заявке
7.	процентная ставка
8.	статус одобрения кредита
9.	сумма кредита в процентах от дохода
10.	не допустил ли заявитель ранее дефолта по кредиту (да, нет)
11.	длина кредитной истории заявителя

**1	Основная часть**

**1.1	Подготовка данных и устранение ошибок**

Первым делом читаем файл с помощью команды read.csv и выводим таблицу.

In [65]:
import pandas as pd

df = pd.read_csv('credit_risk_lab1_v15.csv')
df.head(20)

Id   Age  Income      Home  Emp_length             Intent  Amount   Rate  \
0    0  22.0   59000      RENT       123.0           PERSONAL   35000  16.02   
1    1  21.0    9600       OWN         5.0          EDUCATION    1000  11.14   
2    2  25.0    9600  MORTGAGE         1.0            MEDICAL    5500  12.87   
3    3  23.0   65500      RENT         4.0            MEDICAL   35000  15.23   
4    4  24.0   54400      RENT         8.0            MEDICAL   35000  14.27   
5    5  21.0    9900       OWN         2.0            VENTURE    2500   7.14   
6    6  26.0   77100     RENET         8.0          EDUCATION   35000  12.42   
7    7  24.0   78956      RENT         5.0            MEDICAL   35000  11.11   
8    8  24.0   83000      RENT         8.0           PERSONAL   35000   8.90   
9    9  21.0   10000       OWN         6.0            VENTURE    1600  14.74   
10  10  22.0   85000      RENT         6.0            VENTURE   35000  10.37   
11  11  21.0   10000       OWN         2.0    HOMEIMPROVEMENT    4500   8.63   
12  12  23.0   95000      RENT         2.0            VENTURE   35000   7.90   
13  13  26.0  108160      RENT         4.0          EDUCATION   35000  18.39   
14  14  23.0  115000      RENT         2.0          EDUCATION   35000   7.90   
15  15  23.0  500000  MORTGAGE         7.0  DEBTCONSOLIDATION   30000  10.65   
16  16  23.0  120000      RENT         0.0          EDUCATION   35000   7.90   
17  17  23.0   92111      RENT         7.0            MEDICAL   35000  20.25   
18  18  23.0  113000      RENT         8.0  DEBTCONSOLIDATION   35000  18.25   
19  19  24.0   10800  MORTGAGE         8.0          EDUCATION    1750  10.99   

    Status  Percent_income Default  Cred_length  
0        1            0.59       Y            3  
1        0            0.10       N            2  
2        1            0.57       N            3  
3        1            0.53       N            2  
4        1            0.55       Y            4  
5        1            0.25       N            2  
6        1            0.45       N            3  
7        1            0.44       N            4  
8        1            0.42       N            2  
9        1            0.16       N            3  
10       1            0.41       N            4  
11       1            0.45       N            2  
12       1            0.37       N            2  
13       1            0.32       N            4  
14       0            0.30       N            4  
15       0            0.06       N            3  
16       0            0.29       N            4  
17       1            0.32       N            4  
18       1            0.31       N            4  
19       1            0.16       N            2

Далее необходимо удостовериться в корректности названия столбцов и их соответствия данным варианта. Проверить датасет на наличие дубликатов с помощью метода ```duplicated()```, посчитать и вывести количество дубликатов, а также избавиться от них с помощью метода ```drop_duplicates()```.

In [66]:
duplicated_rows = df.duplicated()
print(duplicated_rows.sum())

2


In [67]:
duplicated_rows = df.duplicated()
df = df.drop_duplicates()
print(duplicated_rows.sum())

2


После удаления строк-дубликатов обновили индексацию используя метод ```reset_index()```, который создает новый датафрейм, в котором индексы исходного датафрейма станут новой колонкой с названием index, а все строки получают обычные индексы без пропусков. В этом методе использовали аргумент ```drop```, чтобы не создавать столбец ```index```.

In [68]:
df = df.drop_duplicates().reset_index(drop=True)
df.head(20)

Id   Age  Income      Home  Emp_length             Intent  Amount   Rate  \
0    0  22.0   59000      RENT       123.0           PERSONAL   35000  16.02   
1    1  21.0    9600       OWN         5.0          EDUCATION    1000  11.14   
2    2  25.0    9600  MORTGAGE         1.0            MEDICAL    5500  12.87   
3    3  23.0   65500      RENT         4.0            MEDICAL   35000  15.23   
4    4  24.0   54400      RENT         8.0            MEDICAL   35000  14.27   
5    5  21.0    9900       OWN         2.0            VENTURE    2500   7.14   
6    6  26.0   77100     RENET         8.0          EDUCATION   35000  12.42   
7    7  24.0   78956      RENT         5.0            MEDICAL   35000  11.11   
8    8  24.0   83000      RENT         8.0           PERSONAL   35000   8.90   
9    9  21.0   10000       OWN         6.0            VENTURE    1600  14.74   
10  10  22.0   85000      RENT         6.0            VENTURE   35000  10.37   
11  11  21.0   10000       OWN         2.0    HOMEIMPROVEMENT    4500   8.63   
12  12  23.0   95000      RENT         2.0            VENTURE   35000   7.90   
13  13  26.0  108160      RENT         4.0          EDUCATION   35000  18.39   
14  14  23.0  115000      RENT         2.0          EDUCATION   35000   7.90   
15  15  23.0  500000  MORTGAGE         7.0  DEBTCONSOLIDATION   30000  10.65   
16  16  23.0  120000      RENT         0.0          EDUCATION   35000   7.90   
17  17  23.0   92111      RENT         7.0            MEDICAL   35000  20.25   
18  18  23.0  113000      RENT         8.0  DEBTCONSOLIDATION   35000  18.25   
19  19  24.0   10800  MORTGAGE         8.0          EDUCATION    1750  10.99   

    Status  Percent_income Default  Cred_length  
0        1            0.59       Y            3  
1        0            0.10       N            2  
2        1            0.57       N            3  
3        1            0.53       N            2  
4        1            0.55       Y            4  
5        1            0.25       N            2  
6        1            0.45       N            3  
7        1            0.44       N            4  
8        1            0.42       N            2  
9        1            0.16       N            3  
10       1            0.41       N            4  
11       1            0.45       N            2  
12       1            0.37       N            2  
13       1            0.32       N            4  
14       0            0.30       N            4  
15       0            0.06       N            3  
16       0            0.29       N            4  
17       1            0.32       N            4  
18       1            0.31       N            4  
19       1            0.16       N            2

Следующим шагом с помощью метода ```info()``` получаем общую информацию о столбцах, а также проверяем датафрейм на наличие неявных дубликатов с помощью метода ```unique()```, который возвращает перечень уникальных значений в столбце, и метода ```value_counts()```, который анализирует столбец и показывает каждое уникальное значение, прия этом подсчитывая частоту его встречаемости в списке.

In [69]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 650 entries, 0 to 649
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Id              650 non-null    int64  
 1   Age             650 non-null    float64
 2   Income          650 non-null    int64  
 3   Home            650 non-null    object 
 4   Emp_length      641 non-null    float64
 5   Intent          650 non-null    object 
 6   Amount          650 non-null    int64  
 7   Rate            584 non-null    float64
 8   Status          650 non-null    int64  
 9   Percent_income  650 non-null    float64
 10  Default         650 non-null    object 
 11  Cred_length     650 non-null    int64  
dtypes: float64(4), int64(5), object(3)
memory usage: 61.1+ KB


In [70]:
print(df['Home'].unique())
print(df['Intent'].unique())
print(df['Default'].unique())

['RENT' 'OWN' 'MORTGAGE' 'RENET' 'OTHER']
['PERSONAL' 'EDUCATION' 'MEDICAL' 'VENTURE' 'HOMEIMPROVEMENT'
 'DEBTCONSOLIDATION']
['Y' 'N' 'No']


In [71]:
df.describe()

Id         Age         Income  Emp_length        Amount  \
count  650.00000  650.000000     650.000000  641.000000    650.000000   
mean   324.50000   24.289231   90104.883077    4.594384  18797.307692   
std    187.78312    7.852809   69615.743919    7.291748   9166.746090   
min      0.00000   21.000000    9600.000000    0.000000   1000.000000   
25%    162.25000   23.000000   44000.000000    2.000000  10000.000000   
50%    324.50000   24.000000   70000.000000    4.000000  22000.000000   
75%    486.75000   25.000000  129497.000000    7.000000  25000.000000   
max    649.00000  144.000000  500000.000000  123.000000  35000.000000   

             Rate      Status  Percent_income  Cred_length  
count  584.000000  650.000000      650.000000   650.000000  
mean    12.305325    0.600000        0.279646     3.012308  
std      3.267110    0.490275        0.150425     0.810088  
min      5.420000    0.000000        0.010000     2.000000  
25%     10.250000    0.000000        0.160000     2.000000  
50%     12.180000    1.000000        0.280000     3.000000  
75%     14.725000    1.000000        0.380000     4.000000  
max     21.210000    1.000000        0.830000     4.000000

В столбцах:

•	'Age' что есть два значения равных 144 и 123, которые необходимо заменить с помощью метода replace() на NaN и 23;

•	'Home' есть значение RENET, которое необходимо заменить на RENT;

•	'Emp_length' есть значения 123 (будет заменено на 23) и nan (будет заменено на 0);

•	'Rate' нет нулевых и отрицательных значений, но есть значения некорректного формата. Необходимо привести все значения к одному формату (float64);

•	'Percent_income' необходимо привести все значения к формату float64;

•	'Default' необходимо заменить значение No на N.

In [72]:
df['Age'] = df['Age'].replace(144, pd.NA)
df['Age'] = df['Age'].replace(123, 23)

df['Home'] = df['Home'].replace('RENET', 'RENT')

df['Emp_length'] = df['Emp_length'].replace(123, 23)
df['Emp_length'] = df['Emp_length'].replace('nan', 0)

df['Default'] = df['Default'].replace('No', 'N')

Столбцы 'Rate' и 'Percent_income' приводим к float64 с помощью метода ```to_numeric()``` с аргументом ```errors='cource'``` (некорректные значения принудительно заменяются на NaN).

In [73]:
df['Rate'] = pd.to_numeric(df['Rate'], errors='coerce')
df['Percent_income'] = pd.to_numeric(df['Percent_income'], errors='coerce')

Далее необходимо посчитать пропуски с помощью метода ```isna()```

In [74]:
df.isna().sum()

Id                 0
Age                2
Income             0
Home               0
Emp_length         9
Intent             0
Amount             0
Rate              66
Status             0
Percent_income     0
Default            0
Cred_length        0
dtype: int64

Удаляем строки с пропущенными значениями с помощью метода ```dropna()```

In [75]:
df = df.dropna()
df.isna().sum()

Id                0
Age               0
Income            0
Home              0
Emp_length        0
Intent            0
Amount            0
Rate              0
Status            0
Percent_income    0
Default           0
Cred_length       0
dtype: int64

**Задание 1**

Группировка по 'Home' и подсчет количества элементов в столбце 'Default'

In [76]:
df.groupby(['Home', 'Default'])['Id'].count()

Home      Default
MORTGAGE  N           87
          Y           26
OTHER     Y            2
OWN       N           71
          Y           12
RENT      N          281
          Y           95
Name: Id, dtype: int64

**Задание 2**

Группировка - статус домовладения (home) по количеству дефолтов (default). Создать датафрейм. Переименовать столбец с количеством в “сount”. Отсортировать по убыванию столбца “count”.

In [77]:
grouped = df.groupby(['Home', 'Default'])['Id'].count().reset_index()
grouped = grouped.rename(columns={'Id': 'count'})
grouped = grouped.sort_values(by='count', ascending=False)
grouped

Home Default  count
5      RENT       N    281
6      RENT       Y     95
0  MORTGAGE       N     87
3       OWN       N     71
1  MORTGAGE       Y     26
4       OWN       Y     12
2     OTHER       Y      2

**Задание 3**

Сводная таблица (pivot_table) - средняя сумма кредита amount для каждой цели кредита. Отсортировать по убыванию. Округлить до трёх знаков.

In [78]:
intent_pivot = df.pivot_table(index='Intent', values='Amount', aggfunc='mean')
intent_pivot = intent_pivot.round(3)
intent_pivot = intent_pivot.sort_values(by='Amount', ascending=False)
intent_pivot

Amount
Intent                      
DEBTCONSOLIDATION  19923.851
EDUCATION          19860.433
MEDICAL            19536.685
PERSONAL           19143.873
VENTURE            18297.196
HOMEIMPROVEMENT    17914.831

**Задание 4**

Сводная таблица (pivot_table) - средняя сумма кредита amount для каждой цели (intent) - строки и возраста - столбцы. Отсортировать по возрастанию intent. Округлить до двух знаков.

In [79]:
intent_age_pivot = df.pivot_table(index='Intent', columns='Age',
                                  values='Amount', aggfunc='mean')
intent_age_pivot = intent_age_pivot.round(2)
intent_age_pivot = intent_age_pivot.sort_index(ascending=True)
intent_age_pivot

Age                    21.0      22.0      23.0      24.0      25.0      26.0
Intent                                                                       
DEBTCONSOLIDATION  10583.33  18285.29  20849.14  20150.00  18603.33  23507.69
EDUCATION          13500.00  20487.50  20580.65  20355.00  19628.41  21758.33
HOMEIMPROVEMENT     4693.75  10216.67   4000.00  18904.76  21266.67  21408.33
MEDICAL            10666.67  18418.75  19323.81  21652.27  19284.00  20011.11
PERSONAL            3633.33  17913.89  18435.87  21205.00  21530.43  18400.00
VENTURE             9910.42  16704.69  22731.67  19826.04  18888.10  18850.00

**Выводы**

Для анализа данных, представленных в лабораторной работе, важно сфокусироваться на результатах группировок и сводных таблиц, чтобы выявить ключевые тенденции в кредитных заявках.

1. Анализ данных по статусу домовладения и количеству дефолтов:

•	Арендаторы (RENT): Среди тех, кто арендует жилье, наблюдается самый высокий уровень дефолтов. Из 377 арендаторов, 95 оказались в дефолте, что составляет около 25%. Это свидетельствует о том, что арендаторы подвергаются большему риску невыплаты кредитов по сравнению с другими группами. Вероятно, это связано с более низкой финансовой стабильностью в данной категории заемщиков.

•	Люди с ипотекой (MORTGAGE): Из 114 заемщиков, находящихся в ипотеке, 26 оказались в дефолте, что составляет примерно 23%. Хотя эта цифра ниже, чем у арендаторов, ипотека все равно несет определенные риски. Финансовые учреждения могут учитывать это при расчете кредитных предложений для людей с ипотекой.

•	Владельцы собственного жилья (OWN): Наименьший процент дефолтов среди владельцев собственного жилья. Из 83 таких заемщиков только 12 (около 14%) оказались в дефолте. Это подтверждает, что владельцы жилья являются наиболее стабильной категорией заемщиков, что связано с их финансовой стабильностью и меньшими рисками невыплаты.

2. Анализ данных по намерениям и возрастам заемщиков:

•	DEBTCONSOLIDATION (Консолидация долгов):

o	В категории заемщиков, намеревающихся консолидировать долги, сумма кредита наиболее стабильна, с небольшими изменениями в зависимости от возраста. Средняя сумма для возрастов от 21 до 26 лет колеблется в пределах 18,823 - 23,579.

•	EDUCATION (Образование):

o	Суммы кредитов на образование варьируются значительно: они растут с возрастом до 20,856 для 24-летних и затем немного уменьшаются.

•	HOMEIMPROVEMENT (Улучшение жилья):

o	Для улучшения жилья средние суммы кредитов растут с возрастом, начиная с 14,746 у 21-летних до 19,268 у 26-летних.

•	MEDICAL (Медицинские расходы):

o	Категории медицинских кредитов суммы держатся на высоком уровне, начиная с 16,665 у 21-летних и достигая пика в 23,371 у 26-летних.

•	PERSONAL (Персональные кредиты):

o	Персональные кредиты характеризуются значительными колебаниями, с наибольшей средней суммой в 21,205 у 24-летних заемщиков.

•	VENTURE (Бизнес):

o	Средняя сумма кредитов для бизнес-проектов растет с 9,910 у 21-летних до 18,888 у 26-летних.

**Общие тенденции:**

В целом, сумма кредитов увеличивается с возрастом в категориях “Debt Consolidation”, “Education”, “Home Improvement” и “Medical”. Максимальные суммы достигаются у заемщиков в возрасте 24-26 лет.

